File: TerminalMacros.java
// TerminalMacros.java
import java.io.*;
import java.nio.file.*;
import java.util.*;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.reflect.TypeToken;

public class TerminalMacros {
    private Map<String, List<String>> macros;
    private String recording;
    private List<String> recordedCmds;
    private String macrosFile;
    private Gson gson;

    public TerminalMacros() {
        macros = new HashMap<>();
        recording = null;
        recordedCmds = new ArrayList<>();
        String home = System.getProperty("user.home");
        macrosFile = home + File.separator + ".terminal_macros.json";
        gson = new GsonBuilder().setPrettyPrinting().create();
        loadMacros();
    }

    private void loadMacros() {
        Path path = Paths.get(macrosFile);
        if (Files.exists(path)) {
            try {
                String json = new String(Files.readAllBytes(path));
                java.lang.reflect.Type type = new TypeToken<Map<String, List<String>>>(){}.getType();
                macros = gson.fromJson(json, type);
                if (macros == null) macros = new HashMap<>();
            } catch (Exception e) {
                macros = new HashMap<>();
            }
        }
    }

    private void saveMacros() {
        try {
            String json = gson.toJson(macros);
            Files.write(Paths.get(macrosFile), json.getBytes());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private int runShell(String cmd) {
        try {
            String shell = "/bin/sh";
            String flag = "-c";
            if (System.getProperty("os.name").toLowerCase().contains("win")) {
                shell = "cmd";
                flag = "/c";
            }
            ProcessBuilder pb = new ProcessBuilder(shell, flag, cmd);
            pb.inheritIO();
            Process p = pb.start();
            return p.waitFor();
        } catch (Exception e) {
            e.printStackTrace();
            return 1;
        }
    }

    public boolean handleCommand(String line) {
        line = line.trim();
        if (line.isEmpty()) return true;
        String[] parts = line.split("\\s+");
        String cmd = parts[0];
        String[] args = parts.length > 1 ? Arrays.copyOfRange(parts, 1, parts.length) : new String[0];

        switch (cmd) {
            case "record":
                if (args.length == 0) {
                    System.out.println("Usage: record <name>");
                    return true;
                }
                String name = args[0];
                if (recording != null) {
                    System.out.println("Already recording macro '" + recording + "'. Use 'stop' to finish.");
                    return true;
                }
                recording = name;
                recordedCmds.clear();
                System.out.println("Recording macro '" + name + "'. Type 'stop' to finish.");
                return true;

            case "stop":
                if (recording == null) {
                    System.out.println("No macro is currently being recorded.");
                    return true;
                }
                if (!recordedCmds.isEmpty()) {
                    macros.put(recording, new ArrayList<>(recordedCmds));
                    saveMacros();
                    System.out.println("Macro '" + recording + "' recorded (" + recordedCmds.size() + " commands).");
                } else {
                    System.out.println("Macro '" + recording + "' is empty, not saved.");
                }
                recording = null;
                recordedCmds.clear();
                return true;

            case "play":
                if (args.length == 0) {
                    System.out.println("Usage: play <name>");
                    return true;
                }
                String macroName = args[0];
                List<String> cmds = macros.get(macroName);
                if (cmds == null) {
                    System.out.println("Macro '" + macroName + "' not found.");
                    return true;
                }
                System.out.println("Executing macro '" + macroName + "'...");
                for (String c : cmds) {
                    System.out.println("> " + c);
                    runShell(c);
                }
                return true;

            case "list":
                if (macros.isEmpty()) {
                    System.out.println("No macros defined.");
                } else {
                    for (Map.Entry<String, List<String>> entry : macros.entrySet()) {
                        System.out.println(entry.getKey() + " (" + entry.getValue().size() + " commands)");
                    }
                }
                return true;

            case "delete":
                if (args.length == 0) {
                    System.out.println("Usage: delete <name>");
                    return true;
                }
                String delName = args[0];
                if (macros.remove(delName) != null) {
                    saveMacros();
                    System.out.println("Macro '" + delName + "' deleted.");
                } else {
                    System.out.println("Macro '" + delName + "' not found.");
                }
                return true;

            case "help":
                System.out.println("Commands:");
                System.out.println("  record <name>   Start recording a macro");
                System.out.println("  stop            Stop recording");
                System.out.println("  play <name>     Execute a macro");
                System.out.println("  list            Show all macros");
                System.out.println("  delete <name>   Delete a macro");
                System.out.println("  help            Show this help");
                System.out.println("  exit            Exit the program");
                System.out.println("Any other command is executed in the shell.");
                return true;

            case "exit":
                return false;

            default:
                if (recording != null) {
                    recordedCmds.add(line);
                    System.out.println("[recording] " + line);
                }
                runShell(line);
                return true;
        }
    }

    public void run() {
        System.out.println("Terminal Macros v1.0. Type 'help' for commands.");
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        while (true) {
            System.out.print("> ");
            try {
                String line = reader.readLine();
                if (line == null) break;
                if (!handleCommand(line)) break;
            } catch (IOException e) {
                break;
            }
        }
        saveMacros();
    }

    public static void main(String[] args) {
        TerminalMacros app = new TerminalMacros();
        app.run();
    }
}

File: terminal_macros.cs
// terminal_macros.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Text;
using System.Text.Json;

class TerminalMacros
{
    private Dictionary<string, List<string>> macros = new Dictionary<string, List<string>>();
    private string recording = null;
    private List<string> recordedCmds = new List<string>();
    private string macrosFile = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.UserProfile), ".terminal_macros.json");

    public TerminalMacros()
    {
        LoadMacros();
    }

    private void LoadMacros()
    {
        if (File.Exists(macrosFile))
        {
            try
            {
                string json = File.ReadAllText(macrosFile);
                macros = JsonSerializer.Deserialize<Dictionary<string, List<string>>>(json) ?? new Dictionary<string, List<string>>();
            }
            catch { macros = new Dictionary<string, List<string>>(); }
        }
    }

    private void SaveMacros()
    {
        string json = JsonSerializer.Serialize(macros, new JsonSerializerOptions { WriteIndented = true });
        File.WriteAllText(macrosFile, json);
    }

    private int RunShell(string cmd)
    {
        ProcessStartInfo psi = new ProcessStartInfo();
        psi.FileName = "/bin/sh";
        psi.Arguments = "-c \"" + cmd.Replace("\"", "\\\"") + "\"";
        psi.RedirectStandardOutput = false;
        psi.RedirectStandardError = false;
        psi.UseShellExecute = false;
        psi.CreateNoWindow = true;
        try
        {
            Process p = Process.Start(psi);
            p.WaitForExit();
            return p.ExitCode;
        }
        catch
        {
            // fallback для Windows
            psi.FileName = "cmd";
            psi.Arguments = "/c " + cmd;
            Process p = Process.Start(psi);
            p.WaitForExit();
            return p.ExitCode;
        }
    }

    public bool HandleCommand(string line)
    {
        line = line.Trim();
        if (string.IsNullOrEmpty(line)) return true;

        string[] parts = line.Split(new char[] { ' ' }, StringSplitOptions.RemoveEmptyEntries);
        string cmd = parts[0];
        string[] args = parts.Length > 1 ? parts[1..] : new string[0];

        switch (cmd)
        {
            case "record":
                if (args.Length == 0)
                {
                    Console.WriteLine("Usage: record <name>");
                    return true;
                }
                string name = args[0];
                if (recording != null)
                {
                    Console.WriteLine($"Already recording macro '{recording}'. Use 'stop' to finish.");
                    return true;
                }
                recording = name;
                recordedCmds.Clear();
                Console.WriteLine($"Recording macro '{name}'. Type 'stop' to finish.");
                return true;

            case "stop":
                if (recording == null)
                {
                    Console.WriteLine("No macro is currently being recorded.");
                    return true;
                }
                if (recordedCmds.Count > 0)
                {
                    macros[recording] = new List<string>(recordedCmds);
                    SaveMacros();
                    Console.WriteLine($"Macro '{recording}' recorded ({recordedCmds.Count} commands).");
                }
                else
                {
                    Console.WriteLine($"Macro '{recording}' is empty, not saved.");
                }
                recording = null;
                recordedCmds.Clear();
                return true;

            case "play":
                if (args.Length == 0)
                {
                    Console.WriteLine("Usage: play <name>");
                    return true;
                }
                string macroName = args[0];
                if (!macros.ContainsKey(macroName))
                {
                    Console.WriteLine($"Macro '{macroName}' not found.");
                    return true;
                }
                Console.WriteLine($"Executing macro '{macroName}'...");
                foreach (string c in macros[macroName])
                {
                    Console.WriteLine($"> {c}");
                    RunShell(c);
                }
                return true;

            case "list":
                if (macros.Count == 0)
                {
                    Console.WriteLine("No macros defined.");
                }
                else
                {
                    foreach (var kv in macros)
                    {
                        Console.WriteLine($"{kv.Key} ({kv.Value.Count} commands)");
                    }
                }
                return true;

            case "delete":
                if (args.Length == 0)
                {
                    Console.WriteLine("Usage: delete <name>");
                    return true;
                }
                string delName = args[0];
                if (macros.Remove(delName))
                {
                    SaveMacros();
                    Console.WriteLine($"Macro '{delName}' deleted.");
                }
                else
                {
                    Console.WriteLine($"Macro '{delName}' not found.");
                }
                return true;

            case "help":
                Console.WriteLine("Commands:");
                Console.WriteLine("  record <name>   Start recording a macro");
                Console.WriteLine("  stop            Stop recording");
                Console.WriteLine("  play <name>     Execute a macro");
                Console.WriteLine("  list            Show all macros");
                Console.WriteLine("  delete <name>   Delete a macro");
                Console.WriteLine("  help            Show this help");
                Console.WriteLine("  exit            Exit the program");
                Console.WriteLine("Any other command is executed in the shell.");
                return true;

            case "exit":
                return false;

            default:
                if (recording != null)
                {
                    recordedCmds.Add(line);
                    Console.WriteLine($"[recording] {line}");
                }
                RunShell(line);
                return true;
        }
    }

    public void Run()
    {
        Console.WriteLine("Terminal Macros v1.0. Type 'help' for commands.");
        while (true)
        {
            Console.Write("> ");
            string line = Console.ReadLine();
            if (line == null) break;
            if (!HandleCommand(line)) break;
        }
        SaveMacros();
    }

    static void Main()
    {
        TerminalMacros app = new TerminalMacros();
        app.Run();
    }
}
